README
NAME
    Module::Install::DOAPChangeSets - write your distribution change log in
    RDF

DESCRIPTION
    This package allows you to write your Changes file in Turtle or RDF/XML
    and autogenerate a human-readable text file.

    To do this, create an RDF file called "Changes.ttl" (or something like
    that) and describe your distribution's changes in RDF using the Dublin
    Core, DOAP, and DOAP Change Sets vocabularies. Then in your Makefile.PL,
    include:

      write_doap_changes "Changes.ttl", "Changes", "turtle";

    This line will read your data from the file named as the first argument,
    parse it using either Turtle or RDFXML parsers (the third argument), and
    output a human-readable changelog to the file named as the second
    argument.

    The defaults are "Changes.ttl", "Changes", "turtle", so if you name the
    files like that, then you can exclude all the arguments and just include
    this in your Makefile.PL:

      write_doap_changes;

    There's also a line you can use to output a Changes.xml file:

      write_doap_changes_xml "Changes.ttl", "Changes.xml", "turtle";

WHY?
    Why not?

BUGS
    Please report any bugs to <http://rt.cpan.org/>.

SEE ALSO
    Module::Install, Module::Install::DOAPChangeSets::Format .

    <http://www.perlrdf.org/>.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENSE
    Copyright (C) 2010 by Toby Inkster

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
############################################################################
## Changes #################################################################
############################################################################

Module-Install-DOAPChangeSets
=============================

Created:      2010-01-08
Home page:    <http://search.cpan.org/dist/Module-Install-DOAPChangeSets/>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Module-Install-DOAPChangeSets>
Maintainer:   Toby Inkster <mailto:tobyink@cpan.org>

0.102 [2011-04-02]
 - Allow change items to have multiple RDF types.
 - Improve sort order of versions in output. Take dates into consideration.

0.101 [2010-07-21]
 - (Bugfix) XML output no longer just an empty file!

0.100 [2010-07-18]
 - Sort change items within each version for text output.
 - Use RDF::Trine::Serializer::RDFXML for XML output.

0.04 [2010-04-07]
 - (Update) Best practice is to use http://purl.org/NET/cpan-uri/.

0.03 [2010-03-26]
 - Use Perl::Version to sort revisions in output.

0.02 [2010-01-27]
 - (Bugfix) Project name was being overridden by URI.

0.01 [2010-01-18]
 - Documentation improvements.

0.00_04 [2010-01-13] # Omega 4
 - Split out RDF::DOAP::ChangeSets as a separate Perl module.

0.00_03 [2010-01-11] # Omega 3
 - Reverse order of versions in Changelog (newest at top)
 - (Addition) Use Text::Wrap to wrap text.
 - (Addition) Support Aaron Cope's changefile vocab as a legacy input
   format.

0.00_02 [2010-01-08] # Omega 2
 - Don't recommend QNames containing '.' -- they're supported by libraptor,
   but illegal in Turtle.

0.00_01 [2010-01-08] # Omega 1



Makefile.PL
use strict;
use warnings;

use inc::Module::Install;

my $dist = 'Module-Install-DOAPChangeSets';
my $fn   = "lib/$dist.pm"; $fn =~ s#-#/#g;

name                $dist;
perl_version_from   $fn;
version_from        $fn;
abstract_from       $fn;
readme_from         $fn;
author              'Toby Inkster <tobyink@cpan.org>';
license             'perl';

test_requires       'Test::More' => '0.61';

requires            'Perl::Version'      => 0;
requires            'RDF::Trine'         => '0.112';
requires            'RDF::Query'         => '2.200';
requires            'File::Slurp'        => 0;
requires            'Text::Wrap'         => 0;
requires            'URI::file'          => 0;
requires            'LWP::Simple'        => 0;

resources(
	'homepage'   => "http://search.cpan.org/dist/$dist/",
	'repository' => "http://goddamn.co.uk/viewvc/perlmods/$dist/",
	'bugtracker' => "http://rt.cpan.org/Dist/Display.html?Queue=$dist",
	);

keywords 'RDF', 'Semantic Web', 'CPAN', 'DOAP', 'Perl', 'Changes';

write_doap_changes;
write_doap_changes_xml;

include 'Test::Signature';
auto_install;
WriteAll(
	'meta' => 1,
	'sign' => 1,
	);


lib/Module/Install/DOAPChangeSets.pm
package Module::Install::DOAPChangeSets;

use 5.008;
use base qw(Module::Install::Base);
use strict;

our $VERSION = '0.102';

sub write_doap_changes {
	my $self = shift;
	$self->admin->write_doap_changes(@_) if $self->is_admin;
}

sub write_doap_changes_xml {
	my $self = shift;
	$self->admin->write_doap_changes_xml(@_) if $self->is_admin;
}

1;

__END__
=head1 NAME

Module::Install::DOAPChangeSets - write your distribution change log in RDF

=head1 DESCRIPTION

This package allows you to write your Changes file in Turtle or RDF/XML and
autogenerate a human-readable text file.

To do this, create an RDF file called "Changes.ttl" (or something like that)
and describe your distribution's changes in RDF using the Dublin Core, DOAP,
and DOAP Change Sets vocabularies. Then in your Makefile.PL, include:

  write_doap_changes "Changes.ttl", "Changes", "turtle";

This line will read your data from the file named as the first argument,
parse it using either Turtle or RDFXML parsers (the third argument), and
output a human-readable changelog to the file named as the second argument.

The defaults are "Changes.ttl", "Changes", "turtle", so if you name the
files like that, then you can exclude all the arguments and just include
this in your Makefile.PL:

  write_doap_changes;

There's also a line you can use to output a Changes.xml file:

  write_doap_changes_xml "Changes.ttl", "Changes.xml", "turtle";

=head1 WHY?

Why not?

=head1 BUGS

Please report any bugs to L<http://rt.cpan.org/>.

=head1 SEE ALSO

L<Module::Install>, L<Module::Install::DOAPChangeSets::Format> .

L<http://www.perlrdf.org/>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 by Toby Inkster

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


lib/RDF/DOAP/ChangeSets.pm
#!/usr/bin/perl

=head1 NAME

RDF::DOAP::ChangeSets - create pretty ChangeLogs from RDF

=head1 SYNOPSIS

 use RDF::DOAP::ChangeSets;
 use URI::file;
 
 my $file     = 'path/to/changelog.rdf';
 my $file_uri = URI::file->new_abs($file);
 
 my $dcs = RDF::DOAP::ChangeSets->new(
             $file_uri, undef, undef, 'RDFXML');
 print $dcs->to_string;

=cut

package RDF::DOAP::ChangeSets;

use 5.008;
use strict;

use File::Slurp qw(slurp);
use LWP::Simple;
use Perl::Version;
use RDF::Trine;
use RDF::Query;
use Text::Wrap;

our $VERSION = '0.102';

=head1 DESCRIPTION

This module takes software changelogs written in the RDF DOAP
Change Sets vocabulary and produces human-readable changelogs.

=over

=item C<< RDF::DOAP::ChangeSets->new($uri, $data, $type, $fmt) >>

Creates and initialises an object.

$uri is a URL for the input data. The URL is used to query the
RDF data for the heading of the output changelog. It may be passed
as either a string, or a L<URI> object.

$data is the RDF data to use as input. It may be passed as a
string, or as an L<RDF::Trine::Model> object. If undefined,
this module will attempt to read data from the URL using
L<LWP::Simple>.

$type gives the constructor a hint as to the RDF vocabulary you
are using. For DOAP Change Sets, use 'current'; for Aaron Cope's
Changefile vocab, use 'legacy'; to autodetect, use 'auto'. By
default, performs autodetection. This module may crash and burn
if you try to mix both vocabs!!

$fmt provides a hint as to what RDF format you're using. By
default, Turtle is assumed. Valid values are whatever
RDF::Trine::Parser->new accepts.

=cut

sub new
{
	my $class = shift;
	my $inuri = shift;
	my $data  = shift;
	my $type  = shift || 'auto';
	
	my $model;
	if (ref $data and $data->isa('RDF::Trine::Model'))
	{
		$model = $data;
	}
	else
	{
		$model = RDF::Trine::Model->new( RDF::Trine::Store->temporary_store );
		
		unless (defined $data)
		{
			if (ref $inuri and $inuri->isa('URI::file'))
			{
				$data = slurp( $inuri->file );
			}
			elsif ($inuri =~ /^(http|file|https|ftp):/i)
			{
				$data = get($inuri);
			}
		}
		
		my $fmt    = shift || 'turtle';
		my $parser = RDF::Trine::Parser->new($fmt);
		$parser->parse_into_model("$inuri", $data, $model);
	}
	
	if (lc $type eq 'auto')
	{
		my $r = RDF::Query->new(
			"ASK WHERE { ?a <http://usefulinc.com/ns/doap#Version> ?b .}")
			->execute($model);
		if ($r->get_boolean)
		{
			$type = 'legacy';
		}
		else
		{
			$type = 'current';
		}
	}	
	
	my $self = { 'model' => $model , 'type' => $type , 'uri' => $inuri } ;
	
	bless $self, $class;
}

=item C<< $changeset->is_legacy >>

Boolean, indicating if a legacy vocab is being used.

=cut

sub is_legacy
{
	my $self = shift;
	return (lc $self->{'type'} eq 'legacy');
}

=item C<< $changeset->is_current >>

Boolean, indicating if the current vocab is being used.

=cut

sub is_current
{
	my $self = shift;
	return !$self->is_legacy(@_);
}

=item C<< $changeset->model >>

RDF::Trine::Model object representing the changelog data.

=cut

sub model
{
	my $self = shift;
	return $self->{'model'};
}

=item C<< $changeset->uri >>

String representing the changelog URI.

=cut

sub uri
{
	my $self = shift;
	return $self->{'uri'} . '';
}

=item C<< $changeset->to_string >>

Creates a human-readable representation of the changelog.

=cut

sub to_string
{
	my $self = shift;
	my $rv = '';
	
	# Get project data for all projects described in the model.
	$self->_project_data;
	
	# Heading
	$rv.= "#" x 76 . "\n";
	$rv.= "## " . $self->{'doctitle'} . " " . ("#" x (72 - length $self->{'doctitle'})) . "\n";
	$rv.= "#" x 76 . "\n\n";
	
	# Create a shortcut to the data.
	my $projects = $self->{'projects'};	

	# foreach project
	foreach my $project (sort keys %$projects)
	{
		# Subheading
		$rv.= $projects->{$project}->{'distname'} . "\n";
		$rv.= ('=' x length $projects->{$project}->{'distname'}) . "\n\n";
		
		# Various interesting data about the project.
		$rv.= sprintf("Created:      %s\n", $projects->{$project}->{'created'})
			if $projects->{$project}->{'created'};
		foreach my $u (sort keys %{ $projects->{$project}->{'homepage'} })
		{
			$rv.= sprintf("Home page:    <%s>\n", $u);
		}
		foreach my $u (sort keys %{ $projects->{$project}->{'bugdatabase'} })
		{
			$rv.= sprintf("Bug tracker:  <%s>\n", $u);
		}
		foreach my $m (sort keys %{ $projects->{$project}->{'maint'} })
		{
			my @mboxes = sort keys %{$projects->{$project}->{'maint'}->{$m}->{'mbox'}};
			my $mbox = shift @mboxes;
			
			if (defined $mbox) { $rv.= sprintf("Maintainer:   %s <%s>\n", $projects->{$project}->{'maint'}->{$m}->{'name'}, $mbox); }
			else               { $rv.= sprintf("Maintainer:   %s\n", $projects->{$project}->{'maint'}->{$m}->{'name'}, $mbox); }
		}
		$rv.= "\n";
		
		# Read in data about this project's releases.
		$self->_release_data($project);
		
		my @revisions = sort {
			if (exists $b->{'issued'} and exists $a->{'issued'})
				{ $b->{'issued'} cmp $a->{'issued'} or Perl::Version->new($b->{'revision'}) cmp Perl::Version->new($a->{'revision'}); }
			else
				{ Perl::Version->new($b->{'revision'}) cmp Perl::Version->new($a->{'revision'}); }
		} values %{$projects->{$project}->{'v'}};
		
		# foreach version
		foreach my $version (@revisions)
		{
			# Version number, release data and version name.
			$rv.= $version->{'revision'};
			$rv.= sprintf(' [%s]', $version->{'issued'})
				if $version->{'issued'};
			$rv.= sprintf(' # %s', $version->{'name'})
				if $version->{'name'};
			$rv.= "\n";
			
			my @changes = sort {
				$a->{type} cmp $b->{type} || $a->{label} cmp $b->{label}
				} values %{$version->{'c'}};
			
			# foreach change
			foreach my $change (@changes)
			{
				my $sigil = '';
				if (defined $change->{'type'} and ref($change->{'type'}) eq 'ARRAY')
				{
					$sigil = join ' ',
						sort
						map { m!doap.changeset.(.+)$!; $1; }
						grep { m!doap.changeset.(.+)$! }
						@{ $change->{'type'} };
					$sigil = "(${sigil}) " if length $sigil;
				}
				# Bullet point
				$rv.= wrap(' - ', '   ', sprintf("%s%s", $sigil, $change->{'label'})) . "\n";
			}
			$rv.= "\n";
		}
		
	}

	return $rv;
}

=item C<< $changeset->to_file($filename) >>

Same as C<to_string>, but outputs to a file.

=cut

sub to_file
{
	my $self = shift;
	my $file = shift;
	
	open OUT, ">$file";
	print OUT $self->to_string;
	close OUT;
}

sub _project_data
{
	my $self = shift;
	my $rv;
	
	if ($self->is_legacy)
	{
		$rv = $self->_project_data__legacy(@_);
	}
	else
	{
		$rv = $self->_project_data__current(@_);
	}

	unless (length $self->{'doctitle'})
	{
		foreach my $project (sort keys %{$self->{'projects'}})
		{
			if (length $self->{'doctitle'} == 0
			or  length $self->{'doctitle'} > $self->{'projects'}->{$project}->{'distname'})
			{
				$self->{'doctitle'} = $self->{'projects'}->{$project}->{'distname'};
			}
		}
		if (length $self->{'doctitle'})
		{
			$self->{'doctitle'} = "Changes for " . $self->{'doctitle'};
		}
		else
		{
			$self->{'doctitle'} = "Changes";
		}
	}
	
	return $rv;
}

sub _project_data__current
{
	my $self  = shift;
	my $model = $self->model;
	my $inuri = $self->uri;

	my $sparql = "
	PREFIX dc: <http://purl.org/dc/terms/>
	PREFIX dcs: <http://ontologi.es/doap-changeset#>
	PREFIX doap: <http://usefulinc.com/ns/doap#>
	PREFIX foaf: <http://xmlns.com/foaf/0.1/>
	PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
	SELECT *
	WHERE
	{
		?project	a doap:Project .
		OPTIONAL { <$inuri> dc:title ?title . }
		OPTIONAL { <$inuri> rdfs:label ?title . }
		OPTIONAL { ?project doap:name ?distname . }
		OPTIONAL { ?project rdfs:label ?distname . }
		OPTIONAL { ?project dc:title ?distname . }
		OPTIONAL { ?project doap:created ?created . }
		OPTIONAL { ?project doap:homepage ?homepage . }
		OPTIONAL { ?project doap:bug-database ?bugdatabase . }
		OPTIONAL
		{
			?project doap:maintainer ?maint .
			?maint foaf:name ?maintname .
			OPTIONAL { ?maint foaf:mbox ?maintmbox . }
		}
	}
	";
	
	my $query    = RDF::Query->new($sparql);
	my $results  = $query->execute($model);
	my $projects = {};
	my $doctitle = '';
	while (my $row = $results->next)
	{
		my $p = $row->{'project'}->as_ntriples;
		$projects->{$p}->{'EXISTS'}++;
		$projects->{$p}->{'distname'} = $row->{'distname'}->literal_value
			if UNIVERSAL::isa($row->{'distname'}, 'RDF::Trine::Node::Literal');
		$projects->{$p}->{'created'} = $row->{'created'}->literal_value
			if UNIVERSAL::isa($row->{'created'}, 'RDF::Trine::Node::Literal');
		$projects->{$p}->{'homepage'}->{ $row->{'homepage'}->uri }++
			if UNIVERSAL::isa($row->{'homepage'}, 'RDF::Trine::Node::Resource');
		$projects->{$p}->{'bugdatabase'}->{ $row->{'bugdatabase'}->uri }++
			if UNIVERSAL::isa($row->{'bugdatabase'}, 'RDF::Trine::Node::Resource');
		$projects->{$p}->{'maint'}->{ $row->{'maint'}->as_ntriples }->{'name'} = $row->{'maintname'}->literal_value
			if UNIVERSAL::isa($row->{'maintname'}, 'RDF::Trine::Node::Literal');
		$projects->{$p}->{'maint'}->{ $row->{'maint'}->as_ntriples }->{'mbox'}->{ $row->{'maintmbox'}->uri }++
			if UNIVERSAL::isa($row->{'maintmbox'}, 'RDF::Trine::Node::Resource');
		$doctitle = $row->{'title'}->literal_value
			if UNIVERSAL::isa($row->{'title'}, 'RDF::Trine::Node::Literal');
	}
	
	foreach my $k (keys %$projects)
	{
		$projects->{$k}->{'distname'} = $k
			unless defined $projects->{$k}->{'distname'};
	}
	
	$self->{'projects'}  = $projects;
	$self->{'doctitle'}  = $doctitle;
}

sub _project_data__legacy
{
	my $self  = shift;
	my $model = $self->model;
	my $inuri = $self->uri;

	my $sparql = "
	PREFIX dc: <http://purl.org/dc/terms/>
	PREFIX doap: <http://usefulinc.com/ns/doap#>
	PREFIX foaf: <http://xmlns.com/foaf/0.1/>
	PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
	SELECT *
	WHERE
	{
		?project	a doap:Project .
		OPTIONAL { <$inuri> dc:title ?title . }
		OPTIONAL { <$inuri> rdfs:label ?title . }
		OPTIONAL { ?project doap:name ?distname . }
		OPTIONAL { ?project rdfs:label ?distname . }
		OPTIONAL { ?project dc:title ?distname . }
		OPTIONAL { ?project doap:created ?created . }
		OPTIONAL { ?project doap:homepage ?homepage . }
		OPTIONAL { ?project doap:bug-database ?bugdatabase . }
		OPTIONAL
		{
			?project doap:maintainer ?maint .
			?maint foaf:name ?maintname .
			OPTIONAL { ?maint foaf:mbox ?maintmbox . }
		}
	}
	";
	
	my $query    = RDF::Query->new($sparql);
	my $results  = $query->execute($model);
	my $projects = {};
	my $doctitle = '';
	while (my $row = $results->next)
	{
		my $p = $row->{'project'}->as_ntriples;
		$projects->{$p}->{'EXISTS'}++;
		$projects->{$p}->{'distname'} = $row->{'distname'}->literal_value
			if UNIVERSAL::isa($row->{'distname'}, 'RDF::Trine::Node::Literal');
		$projects->{$p}->{'created'} = $row->{'created'}->literal_value
			if UNIVERSAL::isa($row->{'created'}, 'RDF::Trine::Node::Literal');
		$projects->{$p}->{'homepage'}->{ $row->{'homepage'}->uri }++
			if UNIVERSAL::isa($row->{'homepage'}, 'RDF::Trine::Node::Resource');
		$projects->{$p}->{'bugdatabase'}->{ $row->{'bugdatabase'}->uri }++
			if UNIVERSAL::isa($row->{'bugdatabase'}, 'RDF::Trine::Node::Resource');
		$projects->{$p}->{'maint'}->{ $row->{'maint'}->as_ntriples }->{'name'} = $row->{'maintname'}->literal_value
			if UNIVERSAL::isa($row->{'maintname'}, 'RDF::Trine::Node::Literal');
		$projects->{$p}->{'maint'}->{ $row->{'maint'}->as_ntriples }->{'mbox'}->{ $row->{'maintmbox'}->uri }++
			if UNIVERSAL::isa($row->{'maintmbox'}, 'RDF::Trine::Node::Resource');
		$doctitle = $row->{'title'}->literal_value
			if UNIVERSAL::isa($row->{'title'}, 'RDF::Trine::Node::Literal');
	}

	foreach my $k (keys %$projects)
	{
		$projects->{$k}->{'distname'} = $k
			unless defined $projects->{$k}->{'distname'};
	}

	$self->{'projects'}  = $projects;
	$self->{'doctitle'}  = $doctitle;
}

sub _release_data
{
	my $self = shift;

	if ($self->is_legacy)
	{
		return $self->_release_data__legacy(@_);
	}
	else
	{
		return $self->_release_data__current(@_);
	}
}

sub _release_data__current
{
	my $self     = shift;
	my $model    = $self->model;
	my $p        = shift;
	my $projects = $self->{'projects'};
	
	my $sparql = "
	PREFIX dc: <http://purl.org/dc/terms/>
	PREFIX dcs: <http://ontologi.es/doap-changeset#>
	PREFIX doap: <http://usefulinc.com/ns/doap#>
	PREFIX foaf: <http://xmlns.com/foaf/0.1/>
	PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
	SELECT *
	WHERE
	{
		$p doap:release ?version .
		?version doap:revision ?revision .
		OPTIONAL { ?version dc:issued ?issued . }
		OPTIONAL { ?version rdfs:label ?vname . }
		OPTIONAL
		{
			?version dcs:changeset [ dcs:item ?item ] .
			OPTIONAL { ?item a ?itemtype . }
			OPTIONAL { ?item rdfs:label ?itemlabel . }
		}
	}
	";

	my $query    = RDF::Query->new($sparql);
	my $results  = $query->execute($model);
	while (my $row = $results->next)
	{
		my $v = $row->{'version'}->as_ntriples;
		$projects->{$p}->{'v'}->{$v}->{'EXISTS'}++;
		
		$projects->{$p}->{'v'}->{$v}->{'revision'} = $row->{'revision'}->literal_value
			if UNIVERSAL::isa($row->{'revision'}, 'RDF::Trine::Node::Literal');
		$projects->{$p}->{'v'}->{$v}->{'issued'} = $row->{'issued'}->literal_value
			if UNIVERSAL::isa($row->{'issued'}, 'RDF::Trine::Node::Literal');
		$projects->{$p}->{'v'}->{$v}->{'name'} = $row->{'vname'}->literal_value
			if UNIVERSAL::isa($row->{'vname'}, 'RDF::Trine::Node::Literal');
		
		if (UNIVERSAL::isa($row->{'item'}, 'RDF::Trine::Node'))
		{
			my $c = $row->{'item'}->as_ntriples;
			$projects->{$p}->{'v'}->{$v}->{'c'}->{$c}->{'label'} = $row->{'itemlabel'}->literal_value
				if UNIVERSAL::isa($row->{'itemlabel'}, 'RDF::Trine::Node::Literal');
			push @{ $projects->{$p}->{'v'}->{$v}->{'c'}->{$c}->{'type'} }, $row->{'itemtype'}->uri
				if UNIVERSAL::isa($row->{'itemtype'}, 'RDF::Trine::Node::Resource')
				and $row->{'itemtype'}->uri ne 'http://ontologi.es/doap-changeset#Change';
		}
	}
}

sub _release_data__legacy
{
	my $self     = shift;
	my $model    = $self->model;
	my $p        = shift;
	my $projects = $self->{'projects'};
	
	my $sparql = "
	PREFIX dc: <http://purl.org/dc/terms/>
	PREFIX asc: <http://aaronland.info/ns/changefile/>
	PREFIX doap: <http://usefulinc.com/ns/doap#>
	PREFIX foaf: <http://xmlns.com/foaf/0.1/>
	PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
	SELECT *
	WHERE
	{
		?version dc:isVersionOf $p .
		?version doap:Version [ doap:revision ?revision ] .
		OPTIONAL { ?version doap:Version [ doap:created ?issued ] . }
		OPTIONAL { ?version rdfs:label ?vname . }
		OPTIONAL { ?version asc:changes [ ?itemtype ?itemlabel ] . }
	}
	";

	my $query    = RDF::Query->new($sparql);
	my $results  = $query->execute($model);
	while (my $row = $results->next)
	{
		my $v = $row->{'version'}->as_ntriples;
		$projects->{$p}->{'v'}->{$v}->{'EXISTS'}++;
		
		$projects->{$p}->{'v'}->{$v}->{'revision'} = $row->{'revision'}->literal_value
			if UNIVERSAL::isa($row->{'revision'}, 'RDF::Trine::Node::Literal');
		$projects->{$p}->{'v'}->{$v}->{'issued'} = $row->{'issued'}->literal_value
			if UNIVERSAL::isa($row->{'issued'}, 'RDF::Trine::Node::Literal');
		$projects->{$p}->{'v'}->{$v}->{'name'} = $row->{'vname'}->literal_value
			if UNIVERSAL::isa($row->{'vname'}, 'RDF::Trine::Node::Literal');
		
		if (UNIVERSAL::isa($row->{'itemlabel'}, 'RDF::Trine::Node'))
		{
			my $c = $row->{'itemlabel'}->as_ntriples;
			$projects->{$p}->{'v'}->{$v}->{'c'}->{$c}->{'label'} = $row->{'itemlabel'}->literal_value
				if UNIVERSAL::isa($row->{'itemlabel'}, 'RDF::Trine::Node::Literal');
				
			if (UNIVERSAL::isa($row->{'itemtype'}, 'RDF::Trine::Node::Resource'))
			{
				my $u = $row->{'itemtype'}->uri;
				
				if ($u =~ m'^http://aaronland.info/ns/changefile/(addition|update|bugfix|removal)$')
				{
					push @{ $projects->{$p}->{'v'}->{$v}->{'c'}->{$c}->{'type'} },
						'http://ontologi.es/doap-changeset#'.(ucfirst $1);
				}
			}
		}
	}	
}

1;

__END__

=back

=head1 BUGS

Please report any bugs to L<http://rt.cpan.org/>.

=head1 SEE ALSO

L<RDF::Trine>, L<Module::Install::DOAPChangeSets>.

L<http://www.perlrdf.org/>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT

Copyright 2010 Toby Inkster

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.
